### PR TITLE
Support async disposal for storage instances, and modify existing clients to use that.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
The new vs cloud cache wants to use IAsyncDisposable, so this modifies our existing system to do the same and updates our own usages to do async dispose calls.

### CODE REVIEW

File: src/Workspaces/Core/Portable/Storage/SQLite/v2/SQLitePersistentStorage.cs [hunk 1]
```diff
@@ -88,14 +88,20 @@ internal partial class SQLitePersistentStorage : AbstractPersistentStorage
         }
 
         public override void Dispose()
+        {
+            var task = DisposeAsync().AsTask();
+            task.Wait();
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Storage/SQLite/v2/SQLitePersistentStorage.cs:93]: this 'Wait' is equivalent to how we disposed before.  Specifically, we would get a flush task and then do a blocking .Wait on that.  This just moves that wait here.  If someone goes through teh DIsposeAsync path, they won't hit this.

File: src/Workspaces/Core/Portable/Storage/SQLite/v2/SQLitePersistentStorage_FlushWrites.cs [hunk 2]
```diff
@@ -59,9 +59,6 @@ private void FlushWritesOnClose()
                     // read/write after releasing us.
                     _shutdownTokenSource.Cancel();
                 }, CancellationToken.None);
-
-            // Wait for that task to finish.
-            writeTask.Wait();
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Storage/SQLite/v2/SQLitePersistentStorage_FlushWrites.cs:64]: we always had a blockign TPL call both before and after this change.  We now have the benefit though that in the DisposeAsync flow, there will be no blocking wait.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/IReferenceCountedDisposable.cs [hunk 3]
```diff
@@ -13,6 +13,9 @@ namespace Roslyn.Utilities
     /// </summary>
     /// <typeparam name="T"></typeparam>
     internal interface IReferenceCountedDisposable<out T> : IDisposable
+#if !CODE_STYLE
+        , IAsyncDisposable
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/IReferenceCountedDisposable.cs:17]: not sure why CODE_STYLE doesn't have access to this type.  seems odd to me.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/ReferenceCountedDisposable.cs [hunk 4]
```diff
@@ -51,7 +53,7 @@ namespace Roslyn.Utilities
     /// even in concurrent execution.</para>
     /// </remarks>
     /// <typeparam name="T">The type of disposable object.</typeparam>
-    internal sealed class ReferenceCountedDisposable<T> : IReferenceCountedDisposable<T>, IDisposable
+    internal sealed class ReferenceCountedDisposable<T> : IReferenceCountedDisposable<T>
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/ReferenceCountedDisposable.cs:56]: IReferenceCountedDisposable is already IDisposable.  So i just removed this as being redundant.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/ReferenceCountedDisposable.cs [hunk 5]
```diff
@@ -51,7 +53,7 @@ namespace Roslyn.Utilities
...
         where T : class, IDisposable
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/ReferenceCountedDisposable.cs:57]: i examined making `T` constrained on IAsyncDisposable as well.  However, this didn't work for all cases of how we use this type.  For examlpe, we make a `ReferenceCountedDisposable<MemoryMappedFile>`.  That inner type is only IDispoable.  So, instead, we require that the tyep be at least IDIsposable, but we respect and forward along IAsyncDisposable if possible.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/RuleSets/VisualStudioRuleSetManager.RuleSetFile.cs [hunk 6]
```diff
@@ -150,6 +152,12 @@ public void Dispose()
                 _disposalCancellationSource.Dispose();
             }
 
+            public ValueTask DisposeAsync()
+            {
+                Dispose();
+                return ValueTaskFactory.CompletedTask;
+            }
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/RuleSets/VisualStudioRuleSetManager.RuleSetFile.cs:159]: ❔ Why are we making this one implement `IAsyncDisposable`?

File: src/Workspaces/Core/Portable/Diagnostics/DiagnosticDataSerializer.cs [hunk 7]
```diff
@@ -49,7 +49,7 @@ public async Task<bool> SerializeAsync(IPersistentStorageService persistentServi
                 WriteDiagnosticData(writer, items, cancellationToken);
             }
 
-            using var storage = await persistentService.GetStorageAsync(project.Solution, cancellationToken).ConfigureAwait(false);
+            await using var storage = await persistentService.GetStorageAsync(project.Solution, cancellationToken).ConfigureAwait(false);
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Diagnostics/DiagnosticDataSerializer.cs:52]: ❔ How does `await using` interact with `ConfigureAwait`?

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/IReferenceCountedDisposable.cs (hunk 3 above)
Comment (sharwell) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/IReferenceCountedDisposable.cs:17]: It's defined in Microsoft.Bcl.AsyncInterfaces. We would have to ship a new assembly in the CodeStyle layer to use this type while still targeting netstandard2.0.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/ReferenceCountedDisposableCache.cs [hunk 8]
```diff
@@ -80,6 +83,12 @@ public void Dispose()
                 // Dispose the underlying value
                 Value.Dispose();
             }
+
+            public ValueTask DisposeAsync()
+            {
+                Dispose();
+                return ValueTaskFactory.CompletedTask;
+            }
```
Comment (sharwell) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/ReferenceCountedDisposableCache.cs:91]: We should remove all of the `DisposeAsync()` methods that just pass through to `Dispose()`. They mislead consumers into believing that code using the types _need_ to be asynchronous for optimal dispose, even though there is no advantage.

**Review by sharwell — CHANGES_REQUESTED** [review]

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/ReferenceCountedDisposable.cs (hunk 4 above)
Comment (sharwell) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/ReferenceCountedDisposable.cs:56]: I felt it was clearer before. `ReferenceCountedDisposable<T>` and `IDisposable` are the primary use cases, and `IReferenceCountedDisposable<T>` is a secondary use case when none of the primary cases works. This change hides a primary interface behind a secondary one.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.OpenFileTracker.cs [hunk 9]
```diff
@@ -442,6 +442,12 @@ public HierarchyEventSink(IVsHierarchy hierarchy, OpenFileTracker openFileTracke
                 void IDisposable.Dispose()
                     => _hierarchy.UnadviseHierarchyEvents(_cookie);
 
+                ValueTask IAsyncDisposable.DisposeAsync()
+                {
+                    ((IDisposable)this).Dispose();
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.OpenFileTracker.cs:447]: >We should remove all of the DisposeAsync() methods that just pass through to Dispose(). They mislead consumers into believing that code using the types need to be asynchronous for optimal dispose, even though there is no advantage.

My thinking was that this helps pay e callers to take the async path. And so if we actually make it async in the future, they're good. Similar to how the dispose path for the sql service is now actually async.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/ReferenceCountedDisposable.cs (hunk 5 above)
Comment (sharwell) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/ReferenceCountedDisposable.cs:57]: This is fine. The alternative would be cloning this to create a `ReferenceCountedAsyncDisposable<T>` class, which is arguably a better match considering `IAsyncDisposable` does not implement `IDisposable`.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.OpenFileTracker.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (sharwell) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.OpenFileTracker.cs:447]: I would prefer to make the changes to the implementation and the call sites at the same time for any given type. We don't save any work by adding more of the implementations in this pull request, so we might as well defer the ones we aren't using now to keep both this change and the future (if/when needed) changes cleaner.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/ReferenceCountedDisposable.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/ReferenceCountedDisposable.cs:56]: ok.  reverted.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/ReferenceCountedDisposable.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/ReferenceCountedDisposable.cs]: yup.  agreed.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/RuleSets/VisualStudioRuleSetManager.RuleSetFile.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/RuleSets/VisualStudioRuleSetManager.RuleSetFile.cs:159]: i'll undo these.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.OpenFileTracker.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.OpenFileTracker.cs:447]: that seems reasonable.

File: src/Workspaces/Core/Portable/Diagnostics/DiagnosticDataSerializer.cs (hunk 7 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Diagnostics/DiagnosticDataSerializer.cs:52]: it seems to be respected: https://devblogs.microsoft.com/dotnet/configureawait-faq/#can-i-use-configureawait-when-await-using-an-iasyncdisposable

File: src/Features/Core/Portable/SolutionCrawler/State/AbstractAnalyzerState.cs [hunk 10]
```diff
@@ -61,7 +61,7 @@ public async Task<TData> TryGetExistingDataAsync(TValue value, CancellationToken
 
             try
             {
-                using var storage = await persistService.GetStorageAsync(solution, cancellationToken).ConfigureAwait(false);
+                await using var storage = await persistService.GetStorageAsync(solution, cancellationToken).ConfigureAwait(false);
```
Comment (sharwell) [inline src/Features/Core/Portable/SolutionCrawler/State/AbstractAnalyzerState.cs:64]: The dispose also needs to be configured in production code, except in cases where JTF is used.

```suggestion
                await using var storage = (await persistService.GetStorageAsync(solution, cancellationToken).ConfigureAwait(false)).ConfigureAwait(false);
```

File: src/Workspaces/Core/Portable/Diagnostics/DiagnosticDataSerializer.cs (hunk 7 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Diagnostics/DiagnosticDataSerializer.cs:52]: Ah, it is respected, but only if configured. Currently we have the equivalent of omitting the `ConfigureAwait` call.

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Diagnostics/DiagnosticDataSerializer.cs]: ok.  have made this change :)

File: src/Workspaces/Core/Portable/Storage/SQLite/v2/SQLitePersistentStorage_FlushWrites.cs (hunk 2 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Storage/SQLite/v2/SQLitePersistentStorage_FlushWrites.cs:64]: Yeah this is good

File: src/Workspaces/Core/Portable/Storage/SQLite/v2/SQLitePersistentStorage.cs [hunk 11]
```diff
@@ -88,14 +88,20 @@ internal partial class SQLitePersistentStorage : AbstractPersistentStorage
...
+        }
+
+        public override async ValueTask DisposeAsync()
         {
             try
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Storage/SQLite/v2/SQLitePersistentStorage.cs:98]: 💡 This method is simplified by starting it with this:

```
// make sure the connection pool is disposed before returning
using var _ = _connectionPool;
```

File: src/Workspaces/Core/Portable/Storage/SQLite/v2/SQLitePersistentStorage.cs (hunk 1 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Storage/SQLite/v2/SQLitePersistentStorage.cs:93]: agree

**Review by sharwell — APPROVED** [review]

**Pull request merged**